runtime/service.go

```diff
@@ -236,6 +240,72 @@ func (s *service) monitorState(ctx context.Context, id, exec_id string, pid uint
 	}
 }
 
+func (s *service) proxyStdio(ctx context.Context, stdin, stdout, stderr string, CID uint32) {
+	go proxyIn(ctx, stdin, CID, internal.StdinPort)
+	go proxyOut(ctx, stdout, CID, internal.StdoutPort)
+	go proxyOut(ctx, stderr, CID, internal.StderrPort)
+}
+
+func proxyIn(ctx context.Context, path string, CID, port uint32) {
+	if path != "" {
+		reader, err := fifo.OpenFifo(ctx, path, syscall.O_RDONLY|syscall.O_NONBLOCK, 0700)
+		if err != nil {
+			log.G(ctx).WithError(err).Error("error opening fifo")
+			return
+		}
+		conn, err := vsock.Dial(CID, port)
+		if err != nil {
+			log.G(ctx).WithError(err).Error("unable to dial agent vsock")
+		}
+		buf := make([]byte, internal.DefaultBufferSize)
+		for {
+			select {
+			case <-ctx.Done():
+				return
+			default:
+				n, err := reader.Read(buf)
+				n, werr := conn.Write(buf[:n])
+				if err != nil {
+					log.G(ctx).WithError(err).Error("error reading from fifo")
+				}
+				if werr != nil {
+					log.G(ctx).WithError(err).Error("error writing to vsock")
+				}
+			}
+		}
+	}
+}
+
+func proxyOut(ctx context.Context, path string, CID, port uint32) {
```

Comment:
`proxyIn` and `proxyOut` are almost identical except a few lines of code? May be have just one function with `direction` parameter?

Reply:
idk if I like the direction parameter but I bet there is a way to extract the shared code part to combine them.